File: kotlin/nimbusflow/src/test/kotlin/com/terminalbench/nimbusflow/CoreTest.kt
package com.terminalbench.nimbusflow

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class CoreTest {
    @Test
    fun planDispatchRespectsCapacity() {
        val out = Allocator.planDispatch(
            listOf(
                DispatchOrder("a", 1, 60),
                DispatchOrder("b", 4, 70),
                DispatchOrder("c", 4, 30)
            ),
            2
        )
        assertEquals(listOf("c", "b"), out.map { it.id })
    }

    @Test
    fun chooseRouteIgnoresBlocked() {
        val route = Routing.chooseRoute(listOf(Route("alpha", 8), Route("beta", 3)), setOf("beta"))
        assertNotNull(route)
        assertEquals("alpha", route!!.channel)
    }

    @Test
    fun nextPolicyEscalates() {
        assertEquals("restricted", Policy.nextPolicy("watch", 3))
    }

    @Test
    fun verifySignatureDigest() {
        val payload = "manifest:v1"
        val digest = Security.digest(payload)
        assertTrue(Security.verifySignature(payload, digest, digest))
        assertFalse(Security.verifySignature(payload, digest.dropLast(1), digest))
    }

    @Test
    fun replayLatestSequenceWins() {
        val replayed = Resilience.replay(listOf(ReplayEvent("x", 1), ReplayEvent("x", 2), ReplayEvent("y", 1)))
        assertEquals(2, replayed.size)
        assertEquals(2, replayed.last().sequence)
    }

    @Test
    fun queueGuardUsesHardLimit() {
        assertFalse(QueueGuard.shouldShed(9, 10, false))
        assertTrue(QueueGuard.shouldShed(11, 10, false))
        assertTrue(QueueGuard.shouldShed(8, 10, true))
    }

    @Test
    fun statisticsPercentileSparse() {
        assertEquals(4, Statistics.percentile(listOf(4, 1, 9, 7), 50))
        assertEquals(0, Statistics.percentile(emptyList(), 90))
    }

    @Test
    fun workflowGraphEnforced() {
        assertTrue(Workflow.canTransition("queued", "allocated"))
        assertFalse(Workflow.canTransition("queued", "arrived"))
    }

    @Test
    fun dispatchOrderUrgencyScore() {
        assertEquals(120, DispatchOrder("d", 3, 30).urgencyScore())
    }

    @Test
    fun flowIntegration() {
        val orders = Allocator.planDispatch(listOf(DispatchOrder("z", 5, 20)), 1)
        val route = Routing.chooseRoute(listOf(Route("north", 4)), emptySet())
        assertEquals(1, orders.size)
        assertNotNull(route)
        assertTrue(Workflow.canTransition("queued", "allocated"))
    }

    @Test
    fun replayConvergence() {
        val a = Resilience.replay(listOf(ReplayEvent("k", 1), ReplayEvent("k", 2)))
        val b = Resilience.replay(listOf(ReplayEvent("k", 2), ReplayEvent("k", 1)))
        assertEquals(a, b)
    }

    @Test
    fun contractsRequiredFields() {
        assertEquals(8160, Contracts.servicePorts["gateway"])
        assertTrue((Contracts.servicePorts["routing"] ?: 0) > 0)
    }
}

File: go/ironfleet/tests/unit/core_test.go
package unit

import (
	"ironfleet/internal/allocator"
	"ironfleet/internal/policy"
	"ironfleet/internal/queue"
	"ironfleet/internal/routing"
	"ironfleet/internal/security"
	"ironfleet/internal/statistics"
	"ironfleet/internal/workflow"
	"ironfleet/pkg/models"
	"testing"
)

func TestPlanDispatchRespectsCapacity(t *testing.T) {
	out := allocator.PlanDispatch([]allocator.Order{
		{ID: "a", Urgency: 1, ETA: "09:30"},
		{ID: "b", Urgency: 3, ETA: "10:00"},
		{ID: "c", Urgency: 3, ETA: "08:30"},
	}, 2)
	if len(out) != 2 || out[0].ID != "c" || out[1].ID != "b" {
		t.Fatalf("unexpected dispatch ordering: %+v", out)
	}
}

func TestChooseRouteIgnoresBlocked(t *testing.T) {
	r := routing.ChooseRoute([]routing.Route{{Channel: "alpha", Latency: 8}, {Channel: "beta", Latency: 2}}, map[string]bool{"beta": true})
	if r == nil || r.Channel != "alpha" {
		t.Fatalf("unexpected route: %+v", r)
	}
}

func TestNextPolicyEscalates(t *testing.T) {
	if got := policy.NextPolicy("watch", 3); got != "restricted" {
		t.Fatalf("expected restricted, got %s", got)
	}
}

func TestVerifySignatureDigest(t *testing.T) {
	sig := security.Digest("manifest:v1")
	if !security.VerifySignature("manifest:v1", sig, sig) {
		t.Fatal("expected valid signature")
	}
	if security.VerifySignature("manifest:v1", sig[:len(sig)-1], sig) {
		t.Fatal("expected invalid signature")
	}
}

func TestShouldShedHardLimit(t *testing.T) {
	if queue.ShouldShed(9, 10, false) {
		t.Fatal("should not shed")
	}
	if !queue.ShouldShed(11, 10, false) {
		t.Fatal("should shed")
	}
	if !queue.ShouldShed(8, 10, true) {
		t.Fatal("should shed under emergency")
	}
}

func TestPercentileSparse(t *testing.T) {
	if got := statistics.Percentile([]int{4, 1, 9, 7}, 50); got != 4 {
		t.Fatalf("expected 4, got %d", got)
	}
	if got := statistics.Percentile(nil, 90); got != 0 {
		t.Fatalf("expected 0, got %d", got)
	}
}

func TestTransitionAllowed(t *testing.T) {
	if !workflow.CanTransition("queued", "allocated") {
		t.Fatal("expected transition")
	}
	if workflow.CanTransition("queued", "arrived") {
		t.Fatal("unexpected transition")
	}
}

func TestDispatchOrderUrgency(t *testing.T) {
	order := models.DispatchOrder{ID: "x", Severity: 3, SLAMinutes: 30}
	if order.UrgencyScore() != 120 {
		t.Fatalf("unexpected urgency: %d", order.UrgencyScore())
	}
}
